fsdev/db.cpp

```diff
@@ -0,0 +1,238 @@
+#include "fs.h"
+#include <sstream>
+
+using namespace std;
+using namespace pas;
+
+string track_column_names[] =
+{
+	// Order must match select / insert code
+	"parent",
+	"fname",
+	"namespace",
+	"artist",
+	"title",
+	"album",
+	"genre",
+	"source",
+	"duration",
+	"publisher",
+	"composer",
+	"track",
+	"copyright",
+	"disc"
+};
+
+string query_columns;
+string parameter_columns;
+
+void InitPreparedStatement()
+{
+	query_columns = "(";
+	parameter_columns = " values (";
+	size_t n = sizeof(track_column_names) / sizeof(string);
+	for (size_t i = 0; i < n; i++)
+	{
+		if (i > 0)
+		{
+			query_columns += ", ";
+			parameter_columns += ", ";
+		}
+		query_columns += track_column_names[i];
+		parameter_columns += "?";
+	}
+	query_columns += ") ";
+	parameter_columns += ") ";
+}
+
+bool TruncateTables(sql::Connection * connection, Logger & _log_)
+{
+	bool rv = false;
+	sql::Statement * stmt = connection->createStatement();
+	try
+	{
+		if (stmt)
+		{
+			stmt->execute("truncate paths;");
+			stmt->execute("truncate tracks;");
+			delete stmt;
+			rv = true;
+		}
+	}
+	catch (sql::SQLException & e)
+	{
+		stringstream ss;
+		ss << "# ERR: SQLException in " << __FILE__;
+		ss << "(" << __FUNCTION__ << ") on line " << __LINE__ << endl;
+		ss << "# ERR: " << e.what();
+		ss << " (MySQL error code: " << e.getErrorCode();
+		ss << ", SQLState: " << e.getSQLState() << " )" << endl;
+		throw LOG2(_log_, ss.str(), LogLevel::FATAL);
+	}
+	return rv;
+}
+
+string SanitizeString(string & s)
+{
+	string rv;
+	rv.reserve(s.size() * 2);
+	for (size_t i = 0; i < s.size(); i++)
+	{
+		switch (s[i])
+		{
+			case '\'':
+			rv += "''";
+			break;
+
+			default:
+			rv += s[i];
+			break;
+		}
+	}
+	return rv;
+}
+
+bool InsertDirectory(sql::Connection * connection, DIRENT * d, string & nspace, Logger & _log_)
+{
+	bool rv = false;
+
+	sql::Statement * stmt = connection->createStatement();
+	try
+	{
+		if (stmt)
```

Comment:
You should probably test `stmt == null` (and throw/return error) outside of this try/catch. It seems more readable to me.